### PR TITLE
FIX: Fixing segmentation fault issue due to double free on SqlHandle::free in Linux

### PR DESCRIPTION
**Pull request by subrata-ms**
> GitHub Issue: #341

-------------------------------------------------------------------
### Summary   

**Problem:** Segmentation fault occurs during Python's garbage collection when freeing ODBC handles.

**Stack Trace Analysis:**

0-2:  Signal handler (SIGSEGV)
3:    libmsodbcsql-18.5.so.1.1 - CRASH LOCATION
4:    SQLFreeHandle() from ODBC driver
5:    SqlHandle::free() from ddbc_bindings
6-11: pybind11 call stack
12:   Python method call
20:   slot_tp_finalize during __del__
21-27: Python GC finalizing objects

**Root Cause:**  The crash occurs when;
1. Python's garbage collector runs during finalization
2. Objects with [__del__] methods are being cleaned up
3. The [__del__] calls [close()] which internally calls SQLFreeHandle()
4. The ODBC driver crashes because:
-     Handle is already freed (double-free)
-     Handle is in invalid state
-     Connection handle freed before cursor statement handle
-     Wrong finalization order due to circular references

**Fix Details:**
This pull request introduces a critical fix to the handle cleanup logic in the SQL bindings for Python, specifically addressing potential segfaults during interpreter shutdown. The change ensures that both statement and database connection handles are not freed if Python is shutting down, preventing invalid memory access.

Handle cleanup logic improvements:

* Updated the `SqlHandle::free()` method in `mssql_python/pybind/ddbc_bindings.cpp` to skip freeing both statement (`SQL_HANDLE_STMT`) and database connection (`SQL_HANDLE_DBC`) handles during Python shutdown, rather than only statement handles. This prevents segfaults caused by freeing handles in the wrong order when their parent resources may have already been released.

### CODE REVIEW

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 1]
```diff
@@ -1127,8 +1127,10 @@ void SqlHandle::free() {
         // CRITICAL FIX: During Python shutdown, don't free STMT handles as
         // their parent DBC may already be freed This prevents segfault when
         // handles are freed in wrong order during interpreter shutdown Type 3 =
```
[the remark this answers is not in the document]
Comment (sumitmsft) [reply mssql_python/pybind/ddbc_bindings.cpp]: This make sense

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 2]
```diff
@@ -1127,8 +1127,10 @@ void SqlHandle::free() {
...
-        // SQL_HANDLE_STMT, Type 2 = SQL_HANDLE_DBC, Type 1 = SQL_HANDLE_ENV
-        if (pythonShuttingDown && _type == 3) {
+        // Type 3 = SQL_HANDLE_STMT (parent: DBC)
+        // Type 2 = SQL_HANDLE_DBC (parent: ENV, which is static and may destruct first)
+        // Type 1 = SQL_HANDLE_ENV (no parent)
+        if (pythonShuttingDown && (_type == 3 || _type == 2)) {
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: Can you confirm if there could be a resource/memory leaks during interpreter shutdown?

So, here we skipped cleanup at shutdown which means any remaining ODBC handles that weren’t explicitly closed (via `.close()` or similar) may not be released properly. This can lead to memory or connection leaks, especially in long-lived or embedded scenarios or stress situations.

Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: Please add documentation explaining that explicit calls to `.close()` are strongly recommended for all connections/cursors, as relying solely on `__del__` may no longer guarantee cleanup under all scenarios. This will help downstream users avoid unexpected resource retention in their scripts, especially in long-running or embedded environments.

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: done

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 2 above)
Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: I have addressed this in latest commit.

Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: This is now handled via "atexit" (Latest commit) during python process shutdown explicitly. All ODBC handles will be explicitly and automatically cleaned up during python process shutdown ( atexit). In Other scenarios ( while python is not shutting down), client/user need to maintain the delete handles in proper sequence i.e                              STMT( type3)->DBC(type2)->Env(type1)

File: tests/test_013_SqlHandle_free_shutdown.py [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+        assert "=== Comprehensive Handle Test ===" in result.stdout
+        assert "Scenario 1: Normal cleanup completed" in result.stdout
+        assert "Scenario 2: Cursor closed, connection left open" in result.stdout
+        assert "Scenario 3: Both cursor and connection left open" in result.stdout
+        assert "Scenario 4: Multiple cursors per connection left open" in result.stdout
+        assert "=== Exiting ===" in result.stdout
+        print(f"PASS: Comprehensive all handle types test")
+
+
+if __name__ == "__main__":
```
Comment (bewithgaurav) [inline tests/test_013_SqlHandle_free_shutdown.py]: this block isn't needed since we have a conftest.py in pytests which take care of setup stuff
if you just run `python -m pytest -v tests\tes_013_....py` you would be able to run this file on its own
I believe we can remove the `__main__` calling of tests.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 4]
```diff
@@ -1124,11 +1124,22 @@ void SqlHandle::free() {
...
+        //
+        // RESOURCE LEAK MITIGATION:
+        // When handles are skipped during shutdown, they are not freed, which could
+        // cause resource leaks. However, this is mitigated by:
+        // 1. Python-side atexit cleanup (in __init__.py) that explicitly closes all
+        //    connections before shutdown, ensuring handles are freed in correct order
+        // 2. OS-level cleanup at process termination recovers any remaining resources
+        // 3. This tradeoff prioritizes crash prevention over resource cleanup, which
+        //    is appropriate since we're already in shutdown sequence
+        if (pythonShuttingDown && (_type == 3 || _type == 2)) {
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: Instead of using 3 and 2: can we define them as contants?

const int SQL_HANDLE_STMT = 3;
const int SQL_HANDLE_DBC = 2;
const int SQL_HANDLE_ENV = 1;

Wouldn't that be a good option?

File: mssql_python/__init__.py [hunk 5]
```diff
@@ -67,6 +69,40 @@
 # Pooling
 from .pooling import PoolingManager
 
+# Global registry for tracking active connections (using weak references)
+_active_connections = weakref.WeakSet()
```
Comment (sumitmsft) [inline mssql_python/__init__.py:74]: As per my understanding, using `WeakSet` allows connections to be garbage collected before `atexit` cleanup runs, which means the segfault still occurs during GC? 
The `atexit` cleanup runs too late and finds nothing?

I can imagine following scenario:
```py
def query():
    conn = connect(...)                           # Added to WeakSet
    return conn.cursor(). fetchall()
    # conn goes out of scope >>  GC collects it
    # SqlHandle::free() called >>  SEGFAULT!
# atexit runs later, finds nothing (weakref died)
```

File: mssql_python/__init__.py [hunk 6]
```diff
@@ -67,6 +69,40 @@
...
+
+def _cleanup_connections():
+    """
+    Cleanup function called by atexit to close all active connections.
+
+    This prevents resource leaks during interpreter shutdown by ensuring
+    all ODBC handles are freed in the correct order before Python finalizes.
+    """
+    # Make a copy of the connections to avoid modification during iteration
+    connections_to_close = list(_active_connections)
```
Comment (sumitmsft) [inline mssql_python/__init__.py]: Seems that `_active_connections` is being modified from multiple threads without synchronization which may cause: RuntimeError or clean up issues or race conditions between registration and cleanup. 

Can you validate it?

File: mssql_python/__init__.py [hunk 7]
```diff
@@ -67,6 +69,40 @@
...
+
+    for conn in connections_to_close:
+        try:
+            # Check if connection is still valid and not closed
+            if hasattr(conn, "_closed") and not conn._closed:
+                # Close will handle both cursors and the connection
+                conn.close()
+        except Exception:
```
Comment (sumitmsft) [inline mssql_python/__init__.py]: Are we purposefully silently ignoring errors during shutdown cleanup? If any issues happen here we will not have any visibility to those issues. Do you not anticipate any issue here?

File: mssql_python/connection.py [hunk 8]
```diff
@@ -239,6 +239,17 @@ def __init__(
         )
         self.setautocommit(autocommit)
 
+        # Register this connection for cleanup before Python shutdown
+        # This ensures ODBC handles are freed in correct order, preventing leaks
+        try:
+            import mssql_python
```
Comment (sumitmsft) [inline mssql_python/connection.py]: This is called for every connection. Isn't this inefficient?

File: mssql_python/connection.py [hunk 9]
```diff
@@ -239,6 +239,17 @@ def __init__(
...
+
+            if hasattr(mssql_python, "_register_connection"):
+                mssql_python._register_connection(self)
+        except (ImportError, AttributeError):
```
Comment (sumitmsft) [inline mssql_python/connection.py]: Silent failure without logging. 

Can we do something like:
```
except (ImportError, AttributeError) as e:
    import warnings ####  Can be on top of connection.py
    warnings.warn(f"Connection registration failed: {e}. Cleanup will rely on __del__", RuntimeWarning)
```

File: mssql_python/__init__.py [hunk 10]
```diff
@@ -67,6 +69,40 @@
...
+    This prevents resource leaks during interpreter shutdown by ensuring
+    all ODBC handles are freed in the correct order before Python finalizes.
+    """
+    # Make a copy of the connections to avoid modification during iteration
+    connections_to_close = list(_active_connections)
+
+    for conn in connections_to_close:
+        try:
+            # Check if connection is still valid and not closed
+            if hasattr(conn, "_closed") and not conn._closed:
```
Comment (sumitmsft) [inline mssql_python/__init__.py:98]: This logic assumed that all connections have `_closed` attribute. Is there any possibility that future connection types don't have this attribute? They'll be silently skipped in that case?

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: Here we assumed python-side atexit cleanup will catch everything (but it might not if connections are created during shutdown). Also, OS will cleanup (true for processes, but not for embedded Python). 

In both these cases, there will definitely be some edge cases where Handle Leak may happen. 

What can be done to tackle these leaked handles?

File: tests/test_013_SqlHandle_free_shutdown.py [hunk 11]
```diff
@@ -0,0 +1,1138 @@
+"""
```
Comment (sumitmsft) [inline tests/test_013_SqlHandle_free_shutdown.py:1]: Can we add a few more test from robustness perspective:

1. Test thread safety when connections are created while shutdown starts. For ex: Thread 1-5 are creating connections, Main thread: Triggers shutdown
Verify: No race conditions, all connections tracked or cleaned

2. Test that _cleanup_connections() can be called multiple times safely. For ex: Create connections, Call _cleanup_connections() multiple times
Verify: No errors, connections cleaned only once

3. Stress Test: Test cleanup performance with many connections. For ex: Create 1000+ connections, Measure cleanup time, Verify: Cleanup completes within reasonable time (shouldn't take forever or even 20 seconds). It should be fairly fast

File: mssql_python/__init__.py (hunk 5 above)
Comment (bewithgaurav) [reply mssql_python/__init__.py:74]: This case should not be a concern, since GC doesn't run when things go out of scope

The scenario describes normal operation (not shutdown), where GC runs
During normal operation, `pythonShuttingDown` is False, so handles are freed properly anyways
The atexit mechanism only matters during actual interpreter shutdown
However, the WeakSet design means connections collected BEFORE shutdown won't be in the set
So, during:

**Normal GC (during runtime):** Works fine, handles freed normally
**Shutdown GC:** WeakSet may miss some connections, but those would be caught by the C++ shutdown protection anyway

@subrata-ms - please add if my analysis here is incorrect
Please also add a small description in comment of why we're using WeakRefs here - and what purpose are they serving (similar to whats added in ddbc bindings)

File: mssql_python/__init__.py (hunk 6 above)
Comment (bewithgaurav) [reply mssql_python/__init__.py]: This is a valid concern
Currently, `_active_connections` is a `WeakSet` that can be accessed from multiple threads without any locks:

**Thread 1** (creating a connection):
```python
# In connection.py
mssql_python._register_connection(self)  
  ↓
# In __init__.py
_active_connections.add(conn)  # ← No lock here!
```

**Thread 2** (also creating a connection at the same time):
```python
_active_connections.add(another_conn)  # ← Same WeakSet, no lock!
```

**Thread 3** (cleanup during shutdown):
```python
# In _cleanup_connections()
connections_to_close = list(_active_connections)  # ← Reading while others write!
```

**Why This Is Dangerous**

`WeakSet` operations in Python are **not thread-safe**. When multiple threads access it simultaneously:

1. **Race condition during `add()`**: Internal data structure corruption
2. **Race condition between iteration and modification**: You might iterate over a set while another thread is adding to it
3. **Potential crashes or data loss**: Missing connections in cleanup, or worse, Python internals corruption

**Real-World Scenario**

```python
# Thread A (web request 1)
conn1 = connect(...)  # Calls _register_connection(conn1)

# Thread B (web request 2) - happens at EXACT same time
conn2 = connect(...)  # Calls _register_connection(conn2)

# Thread C (shutdown signal received)
_cleanup_connections()  # Tries to iterate _active_connections

# Without locks:
# - Connections might be lost
# - Set might be corrupted  
# - Iteration might crash
```

**Potential Solution**

Add a lock to protect all access to `_active_connections`:

```python
import threading

_connections_lock = threading.Lock()
_active_connections = weakref.WeakSet()

def _register_connection(conn):
    """Register a connection for cleanup before shutdown."""
    with _connections_lock:  # ← Only one thread at a time
        _active_connections.add(conn)

def _cleanup_connections():
    """Cleanup function called by atexit to close all active connections."""
    with _connections_lock:  # ← Only one thread at a time
        connections_to_close = list(_active_connections)
    
    # Release lock before cleanup (don't hold lock during slow operations)
    for conn in connections_to_close:
        # ... cleanup code ...
```

Without the lock, even creating connections in different threads could corrupt the tracking set.

**Please also add testcases for the above multi-threaded scenarios**

File: mssql_python/__init__.py [hunk 12]
```diff
@@ -4,8 +4,10 @@
 This module initializes the mssql_python package.
 """
 
+import atexit
```
Comment (bewithgaurav) [inline mssql_python/__init__.py:7]: **Move connection lifecycle management to connection.py**

The current implementation spreads connection lifecycle logic across two files (`__init__.py` and connection.py), requiring a circular import workaround with the lazy `import mssql_python` pattern inside `Connection.__init__()`.

**Recommend:** Move `_active_connections`, `_register_connection()`, and `_cleanup_connections()` to connection.py as module-level code. This would:
- Eliminate the circular dependency pattern
- Keep all connection lifecycle logic in one place
- Reduce bloat in `__init__.py` 
- Make the code more maintainable and easier to understand

The `__init__.py` should only handle module initialization and exports, not resource cleanup logic.

File: mssql_python/connection.py (hunk 8 above)
Comment (bewithgaurav) [reply mssql_python/connection.py]: +1, This is done to bypass circular import using lazy import and should be avoided.
This will be eliminated as a part of moving connection lifecycle management from init to this file

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: fixed.

File: mssql_python/__init__.py (hunk 6 above)
Comment (subrata-ms) [reply mssql_python/__init__.py]: Ya this is a valid scenario. I have fixed the code and added a test for thread safe access to the list _active_connections

File: mssql_python/__init__.py (hunk 7 above)
Comment (subrata-ms) [reply mssql_python/__init__.py]: fixed

File: mssql_python/connection.py (hunk 8 above)
Comment (subrata-ms) [reply mssql_python/connection.py]: fixed. Moved the import statement to module level

File: tests/test_013_SqlHandle_free_shutdown.py (hunk 3 above)
Comment (subrata-ms) [reply tests/test_013_SqlHandle_free_shutdown.py]: fixed.

File: mssql_python/connection.py (hunk 9 above)
Comment (subrata-ms) [reply mssql_python/connection.py]: fixed

File: mssql_python/__init__.py (hunk 10 above)
Comment (subrata-ms) [reply mssql_python/__init__.py:98]: In Connection class (python), "self._closed" is the only attribute that currently denotes the state of the connection (true/false). The whole connection class is currently implemented based on this attribute. Removal of this attribute from this class triggers refactoring of respective functionality under connection class. The change looks good based on current architecture.

File: tests/test_013_SqlHandle_free_shutdown.py (hunk 11 above)
Comment (subrata-ms) [reply tests/test_013_SqlHandle_free_shutdown.py:1]: Added test cases for 1 & 2. For 3, I need to run stress test for connection/connection pool to identify any anomaly.

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: Good point. Our scenario is to protect the double free on ODBC handle. If python process is shutting down and there is connection handle getting created in ODBC and didn't get freed up by "atexit", then it should get cleaned up by garbage collector/parent process cleanup. We will discuss in detail and will cover if any possible edge case need to be added.

File: mssql_python/__init__.py (hunk 12 above)
Comment (gargsaumya) [reply mssql_python/__init__.py:7]: We should avoid calling this as “_active_connections” in naming, because what we are actually tracking here are active ODBC handles that remain open until Python shutdown.

Also, if we move this into connection.py, it gives the impression that the list is tied to individual Connection instances, which isn’t accurate. These are module-level driver resources. With clearer naming (e.g., _active_odbc_handles or _open_driver_handles), it may make more sense to keep this in __init__.py.

@bewithgaurav what do you think of this?

**Review by gargsaumya — CHANGES_REQUESTED** [review]

File: mssql_python/__init__.py (hunk 12 above)
Comment (subrata-ms) [reply mssql_python/__init__.py:7]: Thanks @gargsaumya and @bewithgaurav for the detail review on the code component. However, the design and naming convention seems to be proper and intact. We will have detail discussion on this and will address the outcome with a follow-up PR if needed. My inputs on this would be as below - 
**1.** _active_connections is a global registry/container to hold reference of all the active connections using weak references. Hence we cannot move this inside connection object itself because this container does not belongs to any connection object. Naming convention also seems to be correct as it holds active connections references.

**2.** _register_connection(), and _cleanup_connections() are introduced to operate (kind of getter and setter) on the _active_connections container (list) which is a global entity. Hence it has to be outside of connection class.

File: mssql_python/__init__.py (hunk 5 above)
Comment (subrata-ms) [reply mssql_python/__init__.py:74]: sounds good. Additionally, below is the sequence for python shutdown. Hence if any connections missed in step-1 (current implementation) then it potentially getting take care at step 5.
**Python’s shutdown sequence:**
1. Run atexit callbacks
2. Finalize modules
3. Release GIL
4. Shutdown interpreter
5. Then C++ destructors run

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (sumitmsft) [reply mssql_python/pybind/ddbc_bindings.cpp]: Where was this changed made @subrata-ms ?

Comment (subrata-ms) [reply mssql_python/pybind/ddbc_bindings.cpp]: This already fixed. Somehow it is not reflecting here. Please review it in the ddbc_bindings.cpp and you should see the change as below.
        // RESOURCE LEAK MITIGATION:
        // When handles are skipped during shutdown, they are not freed, which could
        // cause resource leaks. However, this is mitigated by:
        // 1. Python-side atexit cleanup (in __init__.py) that explicitly closes all
        //    connections before shutdown, ensuring handles are freed in correct order
        // 2. OS-level cleanup at process termination recovers any remaining resources
        // 3. This tradeoff prioritizes crash prevention over resource cleanup, which
        //    is appropriate since we're already in shutdown sequence
        **if (pythonShuttingDown && (_type == SQL_HANDLE_STMT || _type == SQL_HANDLE_DBC)) {**
            _handle = nullptr;  // Mark as freed to prevent double-free attempts
            return;
        }

Comment (sumitmsft) [reply mssql_python/pybind/ddbc_bindings.cpp]: Somehow, I was unable to see the change\commit. Now, its visible to me

File: mssql_python/__init__.py (hunk 12 above)
Comment (bewithgaurav) [reply mssql_python/__init__.py:7]: @subrata-ms - understood your points, lets have a detailed discussion on this tomorrow and finalize this part

**Review by sumitmsft — APPROVED** [review]

**Review by bewithgaurav — APPROVED** [review]
